File: admin-portal/src/lib/helpers/actorSelector.ts
/**
 * Copyright 2024 SCOOP Software GmbH, cronn GmbH
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import { ApiAdminActorSelector } from "@eshg/admin-portal-api/serviceDirectory";

import { PartialActorWithId } from "@/lib/components/view/actors/ActorTable";
import { PartialOrgUnitWithId, useOrgUnitsById } from "@/lib/hooks/useOrgUnits";

// eslint-disable-next-line func-style
export const filterOrgUnit =
  (selector: ApiAdminActorSelector) => (orgUnit: PartialOrgUnitWithId) => {
    let result = true;
    if (selector.federalState && orgUnit.federalState) {
      result &&= orgUnit.federalState === selector.federalState;
    }
    if (selector.orgUnitType && orgUnit.type) {
      result &&= orgUnit.type === selector.orgUnitType;
    }
    if (selector.orgUnitName && orgUnit.readableName) {
      result &&= orgUnit.readableName === selector.orgUnitName;
    }
    return result;
  };

export function useFilterActorBySelector(filterActorName = false) {
  const orgUnits = useOrgUnitsById();
  return (selector: ApiAdminActorSelector, actor: PartialActorWithId) => {
    let result = true;
    const orgUnit = actor.orgUnitId ? orgUnits[actor.orgUnitId] : undefined;
    if (orgUnit) {
      result &&= filterOrgUnit(selector)(orgUnit);
    }
    if (selector.actorType && actor.type) {
      result &&= actor.type === selector.actorType;
    }
    if (filterActorName && selector.actorName && actor.readableName) {
      result &&= actor.readableName === selector.actorName;
    }
    return result;
  };
}

File: backend/service-directory/src/main/java/de/eshg/servicedirectory/rule/persistence/entity/ActorSelector.java
/*
 * Copyright 2025 SCOOP Software GmbH, cronn GmbH
 * SPDX-License-Identifier: Apache-2.0
 */

package de.eshg.servicedirectory.rule.persistence.entity;

import de.eshg.lib.common.DataSensitivity;
import de.eshg.lib.common.SensitivityLevel;
import jakarta.persistence.Embeddable;

@Embeddable
@DataSensitivity(SensitivityLevel.PUBLIC)
public record ActorSelector(
    String federalState,
    String orgUnitType,
    String orgUnitName,
    String actorType,
    String actorName) {

  public static ActorSelector empty() {
    return new ActorSelector(null, null, null, null, null);
  }

  @Override
  public String toString() {
    return (federalState == null ? "*" : federalState)
        + "/"
        + (orgUnitType == null ? "*" : orgUnitType)
        + "/"
        + (orgUnitName == null ? "*" : orgUnitName)
        + "/"
        + (actorType == null ? "*" : actorType)
        + "/"
        + (actorName == null ? "*" : actorName);
  }
}
